src/docker/restorableappengine.cc

```diff
@@ -576,32 +576,59 @@ bool RestorableAppEngine::areAppImagesFetched(const App& app) const {
       return false;
     }
 
+    const auto manifest_hash{getContentHash(manifest_file)};
+    if (manifest_hash != manifest_digest.hash()) {
+      LOG_DEBUG << app.name << ": App image manifest hash mismatch; actual: " << manifest_hash
+                << "; expected: " << manifest_digest.hash();
+      return false;
+    }
+
     const auto manifest{Utils::parseJSONFile(blobs_root_ / "sha256" / manifest_digest.hash())};
 
-    std::unordered_set<std::string> image_blobs;
+    // check image config file/blob
+    const auto config_digest{HashedDigest(manifest["config"]["digest"].asString())};
+    const auto config_file{blobs_root_ / "sha256" / config_digest.hash()};
 
-    // add config blob
-    image_blobs.emplace(HashedDigest(manifest["config"]["digest"].asString()).hash());
+    if (!boost::filesystem::exists(config_file)) {
+      LOG_DEBUG << app.name << ": missing App image config file; image: " << image << "; manifest: " << config_file;
+      return false;
+    }
+
+    const auto config_hash{getContentHash(config_file)};
+    if (config_hash != config_digest.hash()) {
+      LOG_DEBUG << app.name << ": App image config hash mismatch; actual: " << config_hash
+                << "; expected: " << config_digest.hash();
+      return false;
+    }
 
-    // add layer blobs
+    // check layers, just check blobs' size since generation of their hashes might consumes
+    // too much CPU for a given device ???
```

Comment:
size is probably fine cause docker is going to verify the hashes when they get imported, right?

Reply:
Unfortunately, `skopeo copy` just fails if one of the existing blobs has incorrect hash. That's why I have to remove the blob size for which  is incorrect (#623).
So, the issue might happen if a size matches while hash is not, to prevent it we need to patch skopeo.